module/VuFind/src/VuFind/Favorites/FavoritesService.php

```diff
@@ -301,6 +312,54 @@ public function save()
         return $this->saveRecordToFavorites(...func_get_args());
     }
 
+    /**
+     * Add/update a resource in the user's account.
+     *
+     * @param UserEntityInterface|int     $userOrId        The user entity or ID saving the favorites
+     * @param ResourceEntityInterface|int $resourceOrId    The resource entity or ID to add/update
+     * @param UserListEntityInterface|int $listOrId        The list entity or ID to store the resource in.
+     * @param array                       $tagArray        An array of tags to associate with the resource.
+     * @param string                      $notes           User notes about the resource.
+     * @param bool                        $replaceExisting Whether to replace all existing tags (true) or
+     * append to the existing list (false).
+     *
+     * @return void
+     */
+    public function saveResourceToFavorites(
+        UserEntityInterface|int $userOrId,
+        ResourceEntityInterface|int $resourceOrId,
+        UserListEntityInterface|int $listOrId,
+        array $tagArray,
+        string $notes,
+        bool $replaceExisting = true
+    ): void {
+        $user = $userOrId instanceof UserEntityInterface
+            ? $userOrId
+            : $this->userService->getUserById($userOrId);
+        $resource = $resourceOrId instanceof ResourceEntityInterface
+            ? $resourceOrId
+            : $this->resourceService->getResourceById($resourceOrId);
+        $list = $listOrId instanceof UserListEntityInterface
+            ? $listOrId
+            : $this->userListService->getUserListById($listOrId);
+
+        // Create the resource link if it doesn't exist and update the notes in any
+        // case:
+        $linkTable = $this->getDbTable('UserResource');
```

Comment:
Database service code in a non-database service?

Reply:
This will be fixed in a subsequent PR; I'm doing the work in steps to keep it from getting too complicated.